### DISCUSSION TITLE
Controlling ESP32 Camera via MQTT

### DISCUSSION THREAD

**Discussion by jlaitinen81** [arendst/Tasmota#11225] [Q&A] [answered]
Hi all! Please help me with my Tasmotized ESP32 Camera. I managed to flash the cam nicely and I can connect to the webUI from my local network just as I should.

I configured the MQTT broker via the webUI and would now like to request a snapshot image via mqtt from the cam. How do I do that? Which mqtt command should I post to get the camera to answer back with an image? 

I’m looking to automate this via Node Red, but can not find the relevant commands.

**Comment by barbudor** [comment, marked as answer]
AFAIK, there is no way to get a single snapshot picture over MQTT
As MQTT message buffer is limited this would not be possible

The possibilities are:
- snapshot picture as `http://ip.addr.of.device/wc.jpg` or `wc.mjpeg` or `snapshot.jpg`
- mjpeg stream as `http://ip.addr.of.device:81/stream`
- optionally rtsp mjpeg stream `rtsp://ip.addr.of.device:8554/mjpeg/1` (using VLC for example)

You can use http request from within node-red

**Reply by sfromis** [reply]
For mjpeg streaming, you should use port 81 instead of the default 80. 
`http://ip.addr.of.device:81/stream`

But indeed, the only thing you can do via MQTT is regular commands and state. No image transfer.

**Reply by jlaitinen81** [reply]
To my understanding MQTT messages can actually be 256Mb in size as max.

But I understand if Tasmota does not support my use case. It's a shame because I can not connect to the remote device's IP directly to get images from the camera. A brokered solution would have been nice.

Have a look at this for example https://medium.com/@chardorn/mqtt-mosquitto-iot-camera-cambia-critters-cef05133c83d

**Reply by barbudor** [reply]
Understand that Tasmota first comes from the ESP8266 with limited RAM. Even on an ESP32 with more RAM, Tasmota is still a general usage firmware/framework and cannot optimized for each use-case, especially something quite specific as a webcam which requires large amounts of RAM. So MQTT buffers of 256kB is just not possible here.
Do not compare a code for the RaspberryPI with a code for an embedded MCU.

You may want to explore a change where the MQTT buffer size is increased to 10KB (as an example) only for the USE_WEBCAM build and implement sending the picture back  in multiple message if it is over the buffer size.

**Reply by jlaitinen81** [reply]
Sure, thanks. As said, the Tasmota Webcam build actually works quite nicely on an ESP32, but I understand now that this is not the best platform for images over MQTT. Will look for something else and use Tasmota for sensors/relays only.

Just wanted to point out that it is not an MQTT limitation, but Tasmota + hardware limitation that images do not work over MQTT as seen from the Raspberry example. There are also ESP32 based MQTT solutions available and I will explore them now that I understand the limitations of Tasmota.
